### PR TITLE
Fixing pip installation

### PR DESCRIPTION
**Pull request by raffienficiaud**
* numpy is required during installation: adding it to "setup_requires"
* getting rid of various warnings for packages not declared explicitly in the "packages" directive
* adding matplotlib

Misc:
* various formatting (black)
* checking the existence of "__package__"
* instructions for pip installations

PS: I can revert the formatting w. black if this makes it easier (or make it consistent for all files)

### CODE REVIEW

File: gwsurrogate/new/spline_evaluation.py [hunk 1]
```diff
@@ -34,68 +34,74 @@
...
-#print("__name__ = %s"%__name__)
-#print("__package__= %s"%__package__)
+# print("__name__ = %s"%__name__)
+# print("__package__= %s"%__package__)
 
-if __package__ is "" or "None": # py2 and py3 compatible 
-  print("setting __package__ to gwsurrogate.new so relative imports work")
-  __package__="gwsurrogate.new"
+if (
+    not "__package__" in locals() or __package__ is None or __package__ == ""
```
Comment (raffienficiaud) [inline gwsurrogate/new/spline_evaluation.py:41]: this is the only relevant change in the file

File: gwsurrogate/new/spline_evaluation.py [hunk 2]
```diff
@@ -34,68 +34,74 @@
...
-#print("__name__ = %s"%__name__)
-#print("__package__= %s"%__package__)
+# print("__name__ = %s"%__name__)
+# print("__package__= %s"%__package__)
 
-if __package__ is "" or "None": # py2 and py3 compatible 
-  print("setting __package__ to gwsurrogate.new so relative imports work")
-  __package__="gwsurrogate.new"
+if (
+    "__package__" not in locals() or __package__ is None or __package__ == ""
```
Comment (raffienficiaud) [inline gwsurrogate/new/spline_evaluation.py:41]: this is the only relevant change in the file

File: setup.py [hunk 3]
```diff
@@ -59,53 +69,66 @@ def run(self):
...
+    long_description_content_type="text/markdown",
+    long_description=long_description,
+    # will start new downloads if these are installed in a non-standard location
+    install_requires=[
+        "numpy",
+        "scipy",
+        "h5py",
+        "scikit-learn",
+        "gwtools",
+        "matplotlib",
```
Comment (raffienficiaud) [inline setup.py:118]: added

File: setup.py [hunk 4]
```diff
@@ -59,53 +69,66 @@ def run(self):
...
+    ],
+    setup_requires=["numpy"],
```
Comment (raffienficiaud) [inline setup.py:120]: added

File: setup.py [hunk 5]
```diff
@@ -1,55 +1,65 @@
...
+    # setup
 
 # To render markdown. See https://github.com/pypa/pypi-legacy/issues/148
 try:
     import pypandoc
-    long_description = pypandoc.convert('README.md', 'rst')
-except ImportError:
-    long_description = open('README.md').read()
+
+    long_description = pypandoc.convert_file("README.md", "rst")
```
Comment (raffienficiaud) [inline setup.py:16]: changed

File: setup.py [hunk 6]
```diff
@@ -59,53 +69,66 @@ def run(self):
...
+    name="gwsurrogate",
+    version=read_main_file("__version__"),
+    author=read_main_file("__author__"),
+    author_email="sfield@umassd.edu",
+    packages=[
+        "gwsurrogate",
+        "gwsurrogate.eval_pysur",
+        "gwsurrogate.new",
+        # "gwsurrogate.precessing_utils",
+        "gwsurrogate.spline_interp_Cwrapper",
```
Comment (raffienficiaud) [inline setup.py:100]: added

File: setup.py [hunk 7]
```diff
@@ -1,55 +1,65 @@
-import sys, os
```
Comment (raffienficiaud) [inline setup.py:1]: removed unused

**Comment by haraldp271** [comment]
@vijayvarma392 -- could you take a look at this?

**Review by duetosymmetry — APPROVED** [review]
LGTM, though I must admit I don't know who uses pyproject.toml or really anything about toml

**Comment by duetosymmetry** [comment]
We need to fix the upstream issue that building the wheel for gwtools is currently failing during the test

**Comment by haraldp271** [comment]
@raffienficiaud is our new parallel programmer at AEI.  Do you know why the build fails?

**Comment by raffienficiaud** [comment]
@haraldp271 nice to meet you here

> We need to fix the upstream issue that building the wheel for gwtools is currently failing during the test

@duetosymmetry @haraldp271 I can have a look

**Comment by raffienficiaud** [comment]
@duetosymmetry the [`setup`](https://bitbucket.org/chadgalley/gwtools/src/master/setup.py) has more or less the same problem as we had here: `numpy` is not available at the time of installation.

The resolution is also more or less the same: we instruct `pip` that `numpy` is required for the installation to proceed. The same `.toml` file should do.

PS: a `toml` is just like a good old `.ini` file, a bit more structured. This is now read by the python package manager, it has the advantage that it is a static description vs. `setup.py` that needs to be executed.

**Comment by raffienficiaud** [comment]
Hi @duetosymmetry and @haraldp271 I am resurrecting the PR after a bit of time, sorry.

I forked and did the same work I did here for `gwtools`, but it turns out that was not the issue in the GH actions. The problem comes from:

1. the `--no-build-isolation` option passed to `pip`
2. the incomatibility w. python 3.12

I don't know the best way to approach this, the compatibility issue will take a bit of time to solve (`imp` module has been removed), for the sake of this PR I am pinning the build to python 3.11 if this is fine for you.
The `--no-build-isolation` I believe is not that standard option and everything works fine without it anyway. All in all, no need to modify `gwtools` for now.

What do you think?

**Comment by raffienficiaud** [comment]
The `imp` module issues apparently come from `nose` itself, see build here: https://github.com/raffienficiaud/gwsurrogate/actions/runs/6698235942/job/18199841787?pr=1

**Comment by raffienficiaud** [comment]
Kind reminder

**Comment by raffienficiaud** [comment]
@vijayvarma392 @duetosymmetry anything I can do that can help clarify the PR and get it merged?

**Comment by vijayvarma392** [comment]
Hi @raffienficiaud, sorry for the long delay with this! Things have been quite hectic :)

@sfield17 mentioned there were some tests he needed to run to make sure the packing was working as expected, and he should be able to get to that once our classes end next week.

**Comment by raffienficiaud** [comment]
Thanks

**Pull request merged**